### PR TITLE
add node native binding

### PR DESCRIPTION
**Pull request by devsnek**

### CODE REVIEW

File: js/index.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  initialize(id, handlers = {}, register = false, steam = '') {
+    binding.initialize(id, handlers, register, steam);
+    immediateLoop();
+  },
+  shutdown() {
+    if (immediate)
+      clearImmediate(immediate);
+    return binding.shutdown();
+  },
+  runCallbacks: binding.runCallbacks,
```
Comment (crmarsh) [inline js/index.js]: You're calling this in your internal loop, no need to expose

File: js/index.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  updateConnection: binding.updateConnection || (() => true),
```
Comment (crmarsh) [inline js/index.js]: I think I would just exclude updateConnection and the DISCORD_DISABLE_IO_THREAD option here. It's meant for a certain kind of performance tweaking that no one using node would be doing (and maybe isn't needed at all).

File: js/index.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+    return binding.updatePresence(presence);
+  },
+  respond(id, reason) {
+    if (typeof reason === 'string') {
+      reason = reason.toUpperCase();
+      if (reason in constants)
+        reason = constants[reason];
+      else if (`REPLY_${reason}` in constants)
+        reason = constants[`REPLY_${reason}`];
```
Comment (crmarsh) [inline js/index.js]: else throw?

File: js/index.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
+const binding = require('bindings')('discord-rpc');
+
+let immediate;
+const hasUpdateConnection = !!binding.updateConnection;
+
+function immediateLoop() {
+  if (hasUpdateConnection)
+    binding.updateConnection();
+  binding.runCallbacks();
+  immediate = setImmediate(immediateLoop);
```
[the remark this answers is not in the document]
Comment (devsnek) [reply js/index.js]: your docs said to call as frequently as possible, but I can turn it down

File: js/index.js (hunk 3 above)
Comment (devsnek) [reply js/index.js]: napi will throw if it doesn't have a number input, no need for me to check

File: js/index.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (crmarsh) [reply js/index.js]: You basically need this to get join/spectate requests in a timely manner (the others are mostly just for debugging) and that stuff happens on more of a human timeframe.

**Comment by msciotti** [comment]
We did this instead https://github.com/discordapp/discord-rpc/commit/b9f9b08606c9988a4ecf31c3e4f844fb86f19ae0.

**Comment by marwanhilmi** [comment]
Sorry to ping closed issue - is there any advantage to using node native binding in this case versus the JS RPC lib? Our initial implementation used native bindings but would be nice to scrap and just use the JS lib.

**Comment by devsnek** [comment]
not particularly. the only benefit I can think of is less overhead (to a point) and since it would be part of this repo it would be kept in sync with the api

**Pull request closed without merging**